[stdlib] Revise `InlineArray` documentation

xwu opened the pull request:
Update documentation for `InlineArray` to include examples with type sugar, reword to avoid certain jargon that may be unfamiliar to some standard library users, and incorporate additional details mostly drawn from the proposal text.

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xwu committed:
Revise `InlineArray` documentation

Update documentation for `InlineArray` to include examples with type sugar, reword to avoid certain jargon that may be unfamiliar to some standard library users, and incorporate additional details mostly drawn from the proposal text.

benrimmington reviewed (commented):
Thank you very much for working on this. I have a few suggestions, but overall you've greatly improved the documentation.

---

Should it be stated that the `[count of Element]` shorthand is preferred?

[Array](<https://developer.apple.com/documentation/swift/array>):

> Shortened forms are preferred

[TSPL — Collection Types](<https://docs.swift.org/swift-book/documentation/the-swift-programming-language/collectiontypes>):

> Although the two forms are functionally identical, the shorthand form is preferred

---

Should the other example code in this file also be updated?

```diff
-  ///     InlineArray<4, Int> { 1 << $0 }  //-> [1, 2, 4, 8]
+  ///     [4 of Int] { 1 << $0 }  //-> [1, 2, 4, 8]
```

```diff
-  ///     InlineArray<4, Int>(first: 1) { $0 << 1 }  //-> [1, 2, 4, 8]
+  ///     [4 of Int](first: 1) { $0 << 1 }  //-> [1, 2, 4, 8]
```

xwu committed:
Update `InlineArray` documentation for reviewer feedback

xwu committed:
[docs] Restore original spelling for invoking `InlineArray` initializers

amartini51 commented on stdlib/public/core/InlineArray.swift:
```diff
@@ -12,34 +12,88 @@
-/// The `InlineArray` type is a specialized array that stores its elements
-/// contiguously inline, rather than allocating an out-of-line region of memory
-/// with copy-on-write optimization.
+/// An `InlineArray` is a specialized container that does not use a separate
```
Use contractions, per API reference style.

```suggestion
/// An `InlineArray` is a specialized container that doesn't use a separate
```

amartini51 commented on stdlib/public/core/InlineArray.swift:
```diff
@@ -12,34 +12,88 @@
-///     MemoryLayout<InlineArray<3, UInt16>>.stride     //-> 6
-///     MemoryLayout<InlineArray<3, UInt16>>.alignment  //-> 2
+/// An `InlineArray` stores its elements contiguously. If an `InlineArray` is a
+/// stored property of a class, then it will be allocated on the heap along with
```
```suggestion
/// stored property of a class, then it's allocated on the heap along with
```

amartini51 reviewed (commented):
Some minor clean-up, but this looks like an improvement to me.

xwu committed:
Update stdlib/public/core/InlineArray.swift

Co-authored-by: Alex Martini <amartini@apple.com>

xwu committed:
Update stdlib/public/core/InlineArray.swift

Co-authored-by: Alex Martini <amartini@apple.com>

xwu committed:
Update stdlib/public/core/InlineArray.swift

Co-authored-by: Alex Martini <amartini@apple.com>

xwu committed:
Update stdlib/public/core/InlineArray.swift

Co-authored-by: Alex Martini <amartini@apple.com>

xwu committed:
Update stdlib/public/core/InlineArray.swift

Co-authored-by: Alex Martini <amartini@apple.com>

xwu commented on stdlib/public/core/InlineArray.swift:
```diff
@@ -12,34 +12,88 @@
-/// with copy-on-write optimization.
+/// An `InlineArray` is a specialized container that does not use a separate
+/// memory allocation just to store its elements. When a value is copied, all of
+/// its elements are copied eagerly. Use an `InlineArray` when you have a fixed
```
Good point. How about: "all of its elements are copied eagerly, like those of a tuple." I fear that bringing up copy-on-write explicitly is going down the rabbit hole too much if someone doesn't know what "eagerly" means, and for someone who does know it'd be extraneous.

amartini51 replied:
I think that's better — it at least gives them something to look up or generalize from.  TSPL doesn't currently have a discussion of this topic for you to cross-reference, but maybe as we expand the discussion of inline arrays there it should get one.

xwu committed:
[docs] Address reviewer feedback on `InlineArray` initialization.

xwu committed:
[docs] Clarify "eagerly" in InlineArray docs

...adjust line wrapping to account for adoption of prior suggestions, and some minor edits.

xwu commented:
@swift-ci please clean test macos platform
